include/fiction/networks/virtual_pi_network.hpp

```diff
@@ -0,0 +1,516 @@
+//
+// Created by benjamin on 14.06.24.
+//
+
+#ifndef FICTION_VIRTUAL_PI_NETWORK_HPP
+#define FICTION_VIRTUAL_PI_NETWORK_HPP
+
+#include "fiction/types.hpp"
+
+#include <mockturtle/algorithms/cleanup.hpp>
+#include <mockturtle/networks/detail/foreach.hpp>
+#include <mockturtle/traits.hpp>
+
+#include <algorithm>
+#include <cassert>
+#include <cstdint>
+#include <cstdlib>
+#include <memory>
+#include <stdexcept>
+#include <type_traits>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include <parallel_hashmap/phmap.h>
+
+namespace fiction
+{
+
+template <typename Ntk>
+struct virtual_storage
+{
+    /**
+     * Shared pointer vector storage for virtual_inputs.
+     */
+    std::vector<typename Ntk::signal> virtual_inputs;
+    /**
+     * Map from virtual_pis to real_pis.
+     */
+    phmap::parallel_flat_hash_map<typename Ntk::signal, typename Ntk::signal> map_virt_to_real_pi;
+};
+
+/* Network with additional "virtual" PIs.
+ *
+ * "Virtual" PIs (Primary Inputs) are used to manage the duplication of PIs in the network. Each "real" PI can have
+ * an arbitrary number of "virtual" PIs, which are copies of the original "real" PI.
+ * A "virtual" PI can be created by duplicating a "real" PI.
+ * To keep track of this relationship, there is a mapping of each "virtual" PI to its corresponding "real" PI in the
+ * network.
+ */
+template <typename Ntk>
+class virtual_pi_network : public Ntk
+{
+  public:
+    using storage = typename Ntk::storage;
+    using node    = typename Ntk::node;
+    using signal  = typename Ntk::signal;
+
+  public:
+    /**
+     * Default constructor for the `virtual_pi_network` class.
+     * Initializes `_storage` as a shared pointer.
+     */
+    explicit virtual_pi_network() : _storage(std::make_shared<virtual_storage<Ntk>>()) {}
+
+    /**
+     * Copy constructor for the `virtual_pi_network` class.
+     * Given a network `ntk`, constructs a new `virtual_pi_network` as a clone of `ntk`.
+     * Initializes `_storage` as a shared pointer.
+     *
+     * @param ntk The network to clone into this object.
+     */
+    explicit virtual_pi_network(const Ntk& ntk) : Ntk(ntk.clone()), _storage(std::make_shared<virtual_storage<Ntk>>())
```

Comment:
Shouldn't we copy the storage as well? What is the purpose of copying only the network?

Reply:
For this purpose we use `clone`. This constructor makes it possible to add the functionalities of a `virtual_pi_network` on top of a network.
`TestType ntk{};

    const auto a = ntk.create_pi();
    const auto b = ntk.create_pi();

    const auto f1 = ntk.create_and(a, b);

    virtual_pi_network vpi{ntk};

    const auto c = vpi.create_virtual_pi(a);`
    
    Tbh this has no real use yet, i may also remove it